3732 Update TTA module based on latest features

Nic-Ma opened the pull request:
Fixes #3732 .

### Description
This PR updated the TTA module with newer MONAI features based on the discussion with @dongyang0122 and @holgerroth .

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

Nic-Ma commented:
/black

Nic-Ma commented:
/build

Nic-Ma commented:
/build

wyli commented on monai/data/test_time_augmentation.py:
```diff
@@ -171,56 +186,23 @@ def __call__(
-        transform_key = InvertibleTransform.trace_key(self.orig_key)
-
-        # create inverter
-        inverter = BatchInverseTransform(self.transform, dl, collate_fn=list_data_collate)
```
so the `BatchInverseTransform` is not used here, anymore is it deprecating?

Nic-Ma replied:
Hi @wyli ,

Thanks for your review.
We already dropped `BatchInverseTransform` in `Invertd` transform and `TransformInverter` handler, and for most cases `num_workers=0` is fastest in `BatchInverseTransform`, but I think it provides the ability to execute extremely huge inverse operations in multi-processing:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/inverse_batch_transform.py#L96
So maybe let's keep it for now?
@rijobro , what do you think? Do you have any other special use cases?

Thanks in advance.

rijobro replied:
I can't think of any other use cases, so perhaps we should keep `BatchInverseTransform`, but use `Invertd` in the vast majority of other cases.

Nic-Ma replied:
@rijobro thanks for your review!

Nic-Ma commented:
/build

rijobro commented on monai/data/test_time_augmentation.py:
```diff
@@ -109,6 +112,9 @@ def __init__(
         meta_key_postfix=DEFAULT_POST_FIX,
+        to_tensor: bool = True,
+        output_device: Union[str, torch.device] = "cpu",
+        post_func: Callable = lambda x: x,
```
```suggestion
        post_func: Callable = _identity,
```

Nic-Ma replied:
Hi @rijobro ,

Sure, updated in the latest commit.

Thanks.

Nic-Ma commented:
/build

Nic-Ma commented:
/build

rijobro commented:
Hi @Nic-Ma, I've created a PR to add to your PR.

1. Improve tests
2. Reduce data type conversion.

The test currently fails, I will try to debug on Monday, but I feel it is important that it works before we merge. We seem to get different results for:

1. numpy and torch inputs, and
2. using the `post_func` instead of `inferrer_fn=post_trans(model(x))`.

Nic-Ma commented:
Hi @rijobro ,

Plan sounds good to me!

Thanks for your help here.

wyli reviewed (approved):
Thanks, it looks good to me.

rijobro commented:
@Nic-Ma my PR is ready: Nic-Ma#362.

Think you will need to merge it for the unit tests to work.

Nic-Ma commented:
> @Nic-Ma my PR is ready: [Nic-Ma#362](https://github.com/Nic-Ma/MONAI/pull/362).
> 
> Think you will need to merge it for the unit tests to work.

Hi @rijobro ,

Thanks for your enhancement, started CI tests in your PR and put some minor comments in it.

Nic-Ma commented:
/black

Nic-Ma commented:
/build

wyli commented:
/build

Nic-Ma commented:
/build

wyli reviewed (approved):
Thanks, it looks good to me. Minor issues (not introduced by this PR) are 1. Runtime error could be a warning to allow for randomness defined in the network such as dropout. 2 np.std has a ddof parameters but this is currently hardcoded cc @rijobro

Nic-Ma commented:
Hi @wyli ,

Thanks for your review.
1. I changed `RuntimeError` to `Warning` and updated the unit tests in the latest commit.
2. About the `std` arg, PyTorch also has another arg `unbiased`:
https://pytorch.org/docs/stable/generated/torch.std.html
I think maybe we don't need to add these additional args so far? If we got some requests, let's add it then?

What do you think?

Thanks.

wyli commented:
/build

wyli commented:
/build
